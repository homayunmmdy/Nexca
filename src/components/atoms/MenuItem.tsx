'use client';
import React from 'react';
import classNames from 'classnames';
import Link from 'next/link';
import { usePathname } from 'next/navigation';
import { HeaderModeType } from '../../types/entities';

interface Props {
   name: string;
   href: string;
   type?: HeaderModeType;
}

const MenuItem = ({ name, href, type = 'desktop' }: Props) => {
   const pathname = usePathname();
   const DesktopClasses = classNames({
      'rounded-xl border-2 border-base-100 hover:bg-indigo-700 hover:text-white transition-colors ease-in-out': true,
      'bg-indigo-700 text-white': pathname == href,
   });

   const MobileClasses = classNames({
      'block w-full border-2 rounded-xl border-indigo-700 text-center p-2': true,
      'bg-indigo-700 text-white': pathname === href,
   });
   return (
      <Link
         href={href}
         title={name}
         className={type === 'desktop' ? DesktopClasses : MobileClasses}
      >
         {name}
      </Link>
   );
};

export default MenuItem;
