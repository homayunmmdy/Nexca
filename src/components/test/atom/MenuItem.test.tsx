import React from 'react';
import { render, screen } from '@testing-library/react';
import { describe, it, expect, vi } from 'vitest';
import MenuItem from '../../atoms/MenuItem';
import { usePathname } from 'next/navigation';

// Mock next/navigation
vi.mock('next/navigation', () => ({
   usePathname: vi.fn(),
}));

// Mock next/link
vi.mock('next/link', () => ({
   default: ({ children, href, title, className }: any) => (
      <a href={href} title={title} className={className}>
         {children}
      </a>
   ),
}));

describe('MenuItem', () => {
   const defaultProps = {
      name: 'Home',
      href: '/home',
   };

   beforeEach(() => {
      vi.mocked(usePathname).mockReturnValue('/');
   });

   it('renders with default props', () => {
      render(<MenuItem {...defaultProps} />);

      const link = screen.getByRole('link', { name: 'Home' });
      expect(link).toBeInTheDocument();
      expect(link).toHaveAttribute('href', '/home');
      expect(link).toHaveAttribute('title', 'Home');
      expect(link).toHaveTextContent('Home');
   });

   it('applies desktop classes by default', () => {
      render(<MenuItem {...defaultProps} />);

      const link = screen.getByRole('link');
      expect(link).toHaveClass(
         'rounded-xl',
         'border-2',
         'border-base-100',
         'hover:bg-indigo-700',
         'hover:text-white',
         'transition-colors',
         'ease-in-out'
      );
   });

   it('applies mobile classes when type is mobile', () => {
      render(<MenuItem {...defaultProps} type="mobile" />);

      const link = screen.getByRole('link');
      expect(link).toHaveClass(
         'block',
         'w-full',
         'border-2',
         'rounded-xl',
         'border-indigo-700',
         'text-center',
         'p-2'
      );
   });

   describe('active state styling', () => {
      it('applies active desktop classes when pathname matches href', () => {
         vi.mocked(usePathname).mockReturnValue('/home');

         render(<MenuItem {...defaultProps} />);

         const link = screen.getByRole('link');
         expect(link).toHaveClass('bg-indigo-700', 'text-white');
      });

      it('does not apply active desktop classes when pathname does not match href', () => {
         vi.mocked(usePathname).mockReturnValue('/about');

         render(<MenuItem {...defaultProps} />);

         const link = screen.getByRole('link');
         expect(link).not.toHaveClass('bg-indigo-700', 'text-white');
      });

      it('applies active mobile classes when pathname matches href', () => {
         vi.mocked(usePathname).mockReturnValue('/home');

         render(<MenuItem {...defaultProps} type="mobile" />);

         const link = screen.getByRole('link');
         expect(link).toHaveClass('bg-indigo-700', 'text-white');
      });

      it('does not apply active mobile classes when pathname does not match href', () => {
         vi.mocked(usePathname).mockReturnValue('/about');

         render(<MenuItem {...defaultProps} type="mobile" />);

         const link = screen.getByRole('link');
         expect(link).not.toHaveClass('bg-indigo-700', 'text-white');
      });
   });

   describe('class combinations', () => {
      it('has correct desktop classes when active', () => {
         vi.mocked(usePathname).mockReturnValue('/home');

         render(<MenuItem {...defaultProps} />);

         const link = screen.getByRole('link');
         const expectedClasses = [
            'rounded-xl',
            'border-2',
            'border-base-100',
            'hover:bg-indigo-700',
            'hover:text-white',
            'transition-colors',
            'ease-in-out',
            'bg-indigo-700',
            'text-white',
         ];

         expectedClasses.forEach((className) => {
            expect(link).toHaveClass(className);
         });
      });

      it('has correct mobile classes when active', () => {
         vi.mocked(usePathname).mockReturnValue('/home');

         render(<MenuItem {...defaultProps} type="mobile" />);

         const link = screen.getByRole('link');
         const expectedClasses = [
            'block',
            'w-full',
            'border-2',
            'rounded-xl',
            'border-indigo-700',
            'text-center',
            'p-2',
            'bg-indigo-700',
            'text-white',
         ];

         expectedClasses.forEach((className) => {
            expect(link).toHaveClass(className);
         });
      });
   });

   it('handles different href and name values', () => {
      const props = {
         name: 'About Us',
         href: '/about',
      };

      render(<MenuItem {...props} />);

      const link = screen.getByRole('link', { name: 'About Us' });
      expect(link).toBeInTheDocument();
      expect(link).toHaveAttribute('href', '/about');
      expect(link).toHaveAttribute('title', 'About Us');
   });

   it('matches snapshot for desktop type', () => {
      vi.mocked(usePathname).mockReturnValue('/');

      const { container } = render(<MenuItem {...defaultProps} />);
      expect(container.firstChild).toMatchSnapshot();
   });

   it('matches snapshot for mobile type', () => {
      vi.mocked(usePathname).mockReturnValue('/');

      const { container } = render(
         <MenuItem {...defaultProps} type="mobile" />
      );
      expect(container.firstChild).toMatchSnapshot();
   });

   it('matches snapshot for active desktop state', () => {
      vi.mocked(usePathname).mockReturnValue('/home');

      const { container } = render(<MenuItem {...defaultProps} />);
      expect(container.firstChild).toMatchSnapshot();
   });
});
