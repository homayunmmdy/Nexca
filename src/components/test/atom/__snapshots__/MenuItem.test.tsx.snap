// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MenuItem > matches snapshot for active desktop state 1`] = `
<a
  class="rounded-xl border-2 border-base-100 hover:bg-indigo-700 hover:text-white transition-colors ease-in-out bg-indigo-700 text-white"
  href="/home"
  title="Home"
>
  Home
</a>
`;

exports[`MenuItem > matches snapshot for desktop type 1`] = `
<a
  class="rounded-xl border-2 border-base-100 hover:bg-indigo-700 hover:text-white transition-colors ease-in-out"
  href="/home"
  title="Home"
>
  Home
</a>
`;

exports[`MenuItem > matches snapshot for mobile type 1`] = `
<a
  class="block w-full border-2 rounded-xl border-indigo-700 text-center p-2"
  href="/home"
  title="Home"
>
  Home
</a>
`;
